### PR TITLE
Allow rendering HTML in shipping package names

### LINKED ISSUE #6903: Shipping package names in checkout block do not render HTML
## Describe the bug

When using the checkout block any modified package names do not render HTML content, instead any HTML is rendered as HTML entities. _This does not occur when using the checkout shortcode._

## To reproduce

1. Create a page using the checkout block.
2. Add the following code to the theme's functions.php file: 

```
add_filter( 'woocommerce_shipping_package_name', 'test_package_name', 1, 3 );
function test_package_name( $title, $index, $package ) {
	return $title . '<br/>test<br/>test2';
}
```

3. Put something in the cart and go to checkout. You will see the HTML entities near the shipping package name. 
4. Replace the checkout block with the checkout shortcode [woocommerce_checkout]
5. View the checkout page. You will see the HTML rendered as expect with HTML line breaks.

## Expected behavior

HTML should be rendered and not converted to HTML entities.

## Screenshots
**Checkout block:**
checkout-block

**Checkout shortcode:** 
checkout-shortcode

### WordPress (please complete the following information):

-   Core version: 6.0.1
-   Gutenberg version (if installed): 
-   WooCommerce version: 6.8.0
-   WooCommerce Blocks version: 8.2.1
-   Site language: English
-   Any other plugins installed: None

### PR DESCRIPTION
**Pull request by nielslange**
Fixes #6903

### Note

Currently, it's not possible to render HTML elements when adding a filter for the shipping package name. This PR aims to enable render HTML when using the shipping package name filter.

Based on @tarhi-saad's [review](https://github.com/woocommerce/woocommerce-blocks/pull/7147#discussion_r977787322), this PR also adds the reusable util `sanitizeHTML`.

#### Cart block

<table>
<tr>
<td valign="top">Before:
<br><br>

Screenshot 2022-09-14 at 12 11 09
</td>
<td valign="top">After:
<br><br>

Screenshot 2022-09-14 at 12 10 23
</td>
</tr>
</table>

#### Checkout block

<table>

<tr>
<td valign="top">Before:
<br><br>

Screenshot 2022-09-14 at 12 11 46
</td>
<td valign="top">After:
<br><br>

Screenshot 2022-09-14 at 12 10 00
</td>
</tr>
</table>

#### Test HTML rendering in shipping package names

1. Install the [Multiple Packages for WooCommerce](https://wordpress.org/plugins/multiple-packages-for-woocommerce/) plugin.
2. Go to `/wp-admin/admin.php?page=wc-settings&tab=multiple_packages`, enable the Multiple Packages for Shipping plugin and select `Product (individual)` for the option `Group By`.
3. Install the [Code Snippets](https://wordpress.org/plugins/code-snippets/) plugin.
4. Add the following code snippet to the site:
```php
add_filter( 'woocommerce_shipping_package_name', 'test_package_name', 1, 3 );
function test_package_name( $title, $index, $package ) {
	return $title . '<br/>test<br/>test2';
}
```
5. Ensure to have one page with the Cart block and one page with the Checkout block.
6. Add two products to the cart.
7. Verify that the page with the Cart block shows the the shipping package name with line breaks instead of showing `Shipping method<br/>test<br/>test2`.
8. Verify that the page with the Checkout block shows the the shipping package name with line breaks instead of showing `Shipping method<br/>test<br/>test2`.
9. Replace the previous code snippet with the following one:
```php
add_filter( 'woocommerce_shipping_package_name', 'test_xss', 1, 3 );
function test_xss( $title, $index, $package ) {
    return $title . '<br><a href="https://woocommerce.com/" target="_blank">WooCommerce.com<img src=x onerror="&#0000106&#0000097&#0000118&#0000097&#0000115&#0000099&#0000114&#0000105&#0000112&#0000116&#0000058&#0000097&#0000108&#0000101&#0000114&#0000116&#0000040&#0000039&#0000088&#0000083&#0000083&#0000039&#0000041"></a>';
}
```
10. Verify that the page with the Cart block and the page with the Checkout block:
  - do not show an XSS alert.
  - do show the package name and the URL [WooCommerce.com](https://woocommerce.com/) in the next line and that the URL opens in a new tab when being clicked.

#### Test reusable sanitise util (only for developer testing)

1. Open `assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx` and change 
```tsx
<div
  className="wc-block-components-shipping-rates-control__package-title"
  dangerouslySetInnerHTML={ sanitizeHTML( packageData.name ) }
/>
```
to
```tsx
<div
  className="wc-block-components-shipping-rates-control__package-title"
  dangerouslySetInnerHTML={ sanitizeHTML( packageData.name, { tags: [ 'a' ] } ) }
/>
```
2. Run `npm run build` to build the extension again.
3. Look up the page with the Cart or the Checkout block from the previous test case.
4. Verify that the the URL [WooCommerce.com](https://woocommerce.com/) is still visible, but that it no longer appears in the next line but in the same line.

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Fix: Render HTML elements on Cart and Checkout blocks when using the `woocommerce_shipping_package_name` filter.

### CODE REVIEW

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (opr) [inline assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: Is it possible for this to open up opportunities for XSS attacks? Can you take a look at #7145 and see how the HTML is limited here? Do you think we should we take the same approach?

**Review by opr — COMMENTED** [review]
Nice work Niels, it works OK but there's a security concern which I left a comment about. Would love to hear your thoughts.

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > Is it possible for this to open up opportunities for XSS attacks? 

Nice catch, @opr! I was able to simulate an XSS attack by adding the following code snippet to the site:
```php
add_filter( 'woocommerce_shipping_package_name', 'test_xss', 1, 3 );
function test_xss( $title, $index, $package ) {
    return '<img src=x onerror="&#0000106&#0000097&#0000118&#0000097&#0000115&#0000099&#0000114&#0000105&#0000112&#0000116&#0000058&#0000097&#0000108&#0000101&#0000114&#0000116&#0000040&#0000039&#0000088&#0000083&#0000083&#0000039&#0000041">';
}
```

> Can you take a look at https://github.com/woocommerce/woocommerce-blocks/pull/7145 and see how the HTML is limited here? Do you think we should we take the same approach?

Thanks for pointing this out. I see that the HTML gets sanitized based on a whitelist with HTML tags and HTML attributes. For consistency reasons, I'd use the same logic in this PR.

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (tarhi-saad) [inline assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: This piece of code will be instrumental! I think we will be using it every time we need to output safe HTML. Could we make it reusable? Maybe it can live in `assets/js/utils` (`@woocommerce/utils`) or `assets/js/base/utils` (`@woocommerce/base-utils`). We can name this file `sanitize-html.ts`, for example! 🙌 

```tsx
const ALLOWED_TAGS = [ 'a', 'b', 'em', 'i', 'strong', 'p', 'br' ];
const ALLOWED_ATTR = [ 'target', 'href', 'rel', 'name', 'download' ];
const sanitizeHTML = ( html: string ) => {
  return {
    __html: sanitize( html, { ALLOWED_TAGS, ALLOWED_ATTR } ),
  };
};
```

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (tarhi-saad) [inline assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: Can we install the `@types/dompurify` package as well? You can install it using this command: `npm i --save-dev @types/dompurify`. This will fix the ESLint error and add TS support!

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (tarhi-saad) [inline assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: @nielslange, why do we need the `__experimentalApplyCheckoutFilter` filter? Are going to run into some issues by using directly `packageData.name` value?

**Review by tarhi-saad — APPROVED** [review]
Hi @nielslange! Thank you for working on this PR! 🙌 I've left some feedback below. Nothing critical, so I approved it, and you can merge it when you're done! 👍

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (hsingyuc) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: @nielslange I think it's [the same code as mine](https://github.com/woocommerce/woocommerce-blocks/pull/7145),  It seems like it's working in this PR, which part do you think is not working?

Comment (hsingyuc) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > I think we will be using it every time we need to output safe HTML. Could we make it reusable? Maybe it can live in assets/js/utils (@woocommerce/utils) or assets/js/base/utils (@woocommerce/base-utils). We can name this file sanitize-html.ts

Sounds good!

@nielslange Since it's working fine, I've merged my PR.  Would you like to move this to the utils file?

Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > @nielslange I think it's [the same code as mine](https://github.com/woocommerce/woocommerce-blocks/pull/7145), It seems like it's working in this PR, which part do you think is not working?

@hsingyuc The default sanitisation is working fine. But looking at the example on https://github.com/cure53/DOMPurify#can-i-configure-dompurify, I see the following example:

```js
// allow only <b> and <q> with style attributes
var clean = DOMPurify.sanitize(dirty, {ALLOWED_TAGS: ['b', 'q'], ALLOWED_ATTR: ['style']});
```

Looking at your implementation, I'd say the code is executed as:

```js
const sanitizeHTML = ( html: string ) => {
  return {
    __html: sanitize( html, { [ 'a', 'b', 'em', 'i', 'strong', 'p', 'br' ],  [ 'target', 'href', 'rel', 'name', 'download' ] } ),
  };
```

As both `[ 'a', 'b', 'em', 'i', 'strong', 'p', 'br' ]` and `[ 'target', 'href', 'rel', 'name', 'download' ]` are unnamed objects, I believe that `DOMPurify.sanitize()` will ignore these values and will execute the code as:

```js
const sanitizeHTML = ( html: string ) => {
  return {
    __html: sanitize( html ),
  };
```

When working on this PR, I changed 
```js
const ALLOWED_TAGS = [ 'a', 'b', 'em', 'i', 'strong', 'p', 'br' ];
```
to
```js
const ALLOWED_TAGS = [ 'b', 'em', 'i', 'strong', 'p', 'br' ];
```
meaning `a` is no longer part of the `ALLOWED_TAGS`. Yet, the anchor element was still visible and did not get removed.

> > I think we will be using it every time we need to output safe HTML. Could we make it reusable? Maybe it can live in assets/js/utils (@woocommerce/utils) or assets/js/base/utils (@woocommerce/base-utils). We can name this file sanitize-html.ts
> 
> Sounds good!
> 
> @nielslange Since it's working fine, I've merged my PR. Would you like to move this to the utils file?

I've created the utils function last Friday, but got sidetracked due to p1663920815156619-slack-C8X6Q7XQU. I'll finalise this PR today, which also moves the sanitisation into a reusable component. I'll let you know once I'm done with that, so that you can use the reusable component as well.

Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > This piece of code will be instrumental! I think we will be using it every time we need to output safe HTML. Could we make it reusable? Maybe it can live in `assets/js/utils` (`@woocommerce/utils`) or `assets/js/base/utils` (`@woocommerce/base-utils`). We can name this file `sanitize-html.ts`, for example! 🙌

Thanks for suggesting this, @tarhi-saad. I just created `sanitizeHTML` as a reusable util and made it available via `@woocommerce/utils`.

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: Thanks for suggesting this. I just added `@types/dompurify` to add TS support for `dompurify`.

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > @nielslange, why do we need the __experimentalApplyCheckoutFilter filter?

I was using `__experimentalApplyCheckoutFilter` as we're using that filter in other files as well. AFAIK, this is the only filter we're currently using for adding filters to the checkout. 

> Are going to run into some issues by using directly packageData.name value?

Would you mind explaining to me what you mean by that, @tarhi-saad?

Comment (tarhi-saad) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > Would you mind explaining to me what you mean by that, @tarhi-saad?

@nielslange, I was just wondering why we need to apply a filter to get the package name in this instance, knowing that on the initial code we didn't have to:

1. Previous code (getting package name directly from the `packageData` props):

```jsx
<div className="wc-block-components-shipping-rates-control__package-title">
  { packageData.name } // ==> This line
</div>
```

2. Current code (Getting the package name through a filter):

```jsx
const data = __experimentalApplyCheckoutFilter( { // ===> The filter
  filterName: 'itemName',
  defaultValue: packageData.name,
} );

<div
  className="wc-block-components-shipping-rates-control__package-title"
  dangerouslySetInnerHTML={ { __html: data } } // ==> The same package name
/>
```

Is there a specific reason why we are using the `__experimentalApplyCheckoutFilter` filter?

Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: Ah, that's what you mean. I used the `__experimentalApplyCheckoutFilter` filter based on @opr [comment](https://github.com/woocommerce/woocommerce-blocks/issues/6903#issuecomment-1227105090). I should have double-checked why we want to run the input through a filter. 

@opr Would you mind clarifying what you had in mind [in this comment](https://github.com/woocommerce/woocommerce-blocks/issues/6903#issuecomment-1227105090)? Did you intend to track the usage of the checkout filter by running the input through a filter?

Comment (opr) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: [The `itemName` filter](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md#cart-line-items) is used to modify the item name to make it different in the Cart/Checkout blocks than it is in the back end. This was required by 3991-gh-woocommerce/woocommerce-subscriptions and added in #3807

What is `packageData.name` here, is it the item name? If not then it should not go through that specific filter, only item names should.

Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > [The `itemName` filter](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/third-party-developers/extensibility/checkout-block/available-filters.md#cart-line-items) is used to modify the item name to make it different in the Cart/Checkout blocks than it is in the back end. This was required by 3991-gh-woocommerce/woocommerce-subscriptions and added in #3807
> 
> What is `packageData.name` here, is it the item name? If not then it should not go through that specific filter, only item names should.

Thanks for clarifying this, @opr. `packageData.name` is not the name of the cart item, but the name of the shipping package. Therefore, I removed the `__experimentalApplyCheckoutFilter`.

@tarhi-saad Goo that you caught this one! Would you mind doing a final review of this PR?

**Review by tarhi-saad — APPROVED** [review]
Thank you, @nielslange, for the updates! I tested it and everything is working fine. I only updated the code snippets for `developer testing` in the PR description:

```jsx
<div
  className="wc-block-components-shipping-rates-control__package-title"
  dangerouslySetInnerHTML={ sanitizeHTML( packageData.name, { tags: [ 'a' ] } ) }
/>
```

So, it looks good to me! Let's ship it! 🚀

**Comment by nielslange** [comment]
> Thank you, @nielslange, for the updates! I tested it and everything is working fine. I only updated the code snippets for `developer testing` in the PR description:
> 
> ```js
> <div
>   className="wc-block-components-shipping-rates-control__package-title"
>   dangerouslySetInnerHTML={ sanitizeHTML( packageData.name, { tags: [ 'a' ] } ) }
> />
> ```
> 
> So, it looks good to me! Let's ship it! 🚀

Awesome! Thanks for updating the instructions, @saad! 🙏

File: assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx
Comment (hsingyuc) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]:
```
const sanitizeHTML = ( html: string ) => {
  return {
    __html: sanitize( html, { [ 'a', 'b', 'em', 'i', 'strong', 'p', 'br' ],  [ 'target', 'href', 'rel', 'name', 'download' ] } ),
  };
```

@nielslange I don't know if you saw my [previous comment](https://github.com/woocommerce/woocommerce-blocks/pull/7145#discussion_r978890021) but this isn't what happens.  This would throw an error because an object can't be created without keys.  For example, try this in your browser:

```
const myProperty = ['a', 'b', 'c' ];
const myObj = { myProperty };
console.log( myObj );
```

This will output:

```
{
    "myProperty": [ "a", "b", "c" ]
}
```

So I suspect something else was happening in your testing.  Maybe the files weren't rebuilt after you removed the anchor tag.

>Would you like to move this to the utils file?

It looks like you've merged this code.  Is there any reason you did not move my method to the utils file?  Or should I open an issue to clean this up so we only have one sanitize method?

Comment (nielslange) [reply assets/js/base/components/cart-checkout/shipping-rates-control-package/index.tsx]: > @nielslange I don't know if you saw my [previous comment](https://github.com/woocommerce/woocommerce-blocks/pull/7145#discussion_r978890021) but this isn't what happens. This would throw an error because an object can't be created without keys. For example, try this in your browser:
> 
> ```
> const myProperty = ['a', 'b', 'c' ];
> const myObj = { myProperty };
> console.log( myObj );
> ```
> 
> This will output:
> 
> ```
> {
>     "myProperty": [ "a", "b", "c" ]
> }
> ```
> 
> So I suspect something else was happening in your testing. Maybe the files weren't rebuilt after you removed the anchor tag.
> 
> >

I've seen your comment, but replied here to avoid that we're discussing the same topic on two different places. Thanks for adding more context and showing the example. Apparently, something went wrong on my end, and I was able to reproduce your result.

> Or should I open an issue to clean this up so we only have one sanitize method?

I see that you already created the PR https://github.com/woocommerce/woocommerce-blocks/pull/7231, so I'll leave this up to you.

**Pull request merged**
